File: python/with_java/programmers/level 2/롤케이크 자르기.py
def solution(topping):
    answer = 0
    cnt = len(set(topping))
    d = {}
    for t in topping:
        d[t] = d.get(t, 0) + 1
    s = set()
    while topping and cnt >= len(s):
        t = topping.pop()
        d[t] -= 1
        if d[t] == 0:
            cnt -= 1
        s.add(t)
        if cnt == len(s):
            answer += 1
        
    return answer

File: java/src/main/java/ps_traning/programmers/level_2/롤케이크_자르기.java
package ps_traning.programmers.level_2;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class 롤케이크_자르기 {
    public int solution(int[] topping) {
        int answer = 0;
        Map<Integer, Integer> map = new HashMap<>();
        for (int t : topping) map.put(t, map.getOrDefault(t, 0) + 1);
        int cnt = map.size();
        Set<Integer> set = new HashSet<>();
        for (int t : topping) {
            if (cnt < set.size()) break;
            map.replace(t, map.get(t) - 1);
            if (map.get(t) == 0) cnt--;
            set.add(t);
            if (cnt == set.size()) answer++;
        }


        return answer;
    }
}
